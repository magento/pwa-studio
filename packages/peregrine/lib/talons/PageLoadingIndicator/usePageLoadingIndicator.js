import { useEffect, useRef, useState } from 'react';
import { useAppContext } from '../../context/app';

const GROUP_LOADING_DELAY = 750;

export default () => {
    const [{ isPageLoading }] = useAppContext();
    const doneTimeoutRef = useRef();
    const [loadingState, setLoadingState] = useState('off');

    useEffect(() => {
        if (isPageLoading) {
            setLoadingState('loading');
        } else if (
            typeof clearTimeout !== 'undefined' &&
            typeof setTimeout !== 'undefined'
        ) {
            setLoadingState('done');

            if (doneTimeoutRef && doneTimeoutRef.current !== null) {
                clearTimeout(doneTimeoutRef.current);
            }

            doneTimeoutRef.current = setTimeout(() => {
                setLoadingState(currentLoadingState => {
                    if (currentLoadingState === 'loading') {
                        return 'loading';
                    }

                    return 'off';
                });
            }, GROUP_LOADING_DELAY);
        } else {
            setLoadingState('off');
        }

        return () => {
            if (
                typeof clearTimeout !== 'undefined' &&
                doneTimeoutRef &&
                doneTimeoutRef.current !== null
            ) {
                clearTimeout(doneTimeoutRef.current);
            }
        };
    }, [isPageLoading, doneTimeoutRef]);

    return {
        isPageLoading,
        loadingState
    };
};
