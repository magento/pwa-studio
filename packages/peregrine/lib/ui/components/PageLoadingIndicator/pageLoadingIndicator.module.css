.root {
    composes: bg-gray-100 from global;
    composes: h-[2px] from global;
    composes: relative from global;
    composes: overflow-hidden from global;
    composes: w-full from global;
}

.root_absolute {
    composes: root;

    composes: absolute from global;
    composes: bottom-0 from global;
    composes: left-0 from global;
}

.indicator_off {
    composes: absolute from global;
    composes: bg-brand-dark from global;
    composes: h-full from global;
    composes: left-0 from global;
    composes: opacity-0 from global;
    composes: top-0 from global;
    composes: w-full from global;
    transform: translateX(-100%);
    transform-origin: left;
    transition: transform 0.25s linear;
}

.indicator_loading {
    composes: indicator_off;
    composes: opacity-100 from global;
    transform: translateX(-25%);
}

.indicator_done {
    composes: indicator_off;
    composes: opacity-100 from global;
    transform: translateX(0%);
}
