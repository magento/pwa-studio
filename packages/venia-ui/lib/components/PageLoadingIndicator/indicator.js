import React from 'react';

import defaultClasses from './indicator.css';
import { useStyle } from '../../classify';
import { RotateCw as LoaderIcon } from 'react-feather';
import Icon from '../Icon';

const PageLoadingIndicator = props => {
    const classes = useStyle(defaultClasses, props.classes);

    return (
        <div className={classes.root}>
            <Icon
                src={LoaderIcon}
                size={24}
                classes={{ root: classes.indicator }}
            />
        </div>
    );
};

export default PageLoadingIndicator;
