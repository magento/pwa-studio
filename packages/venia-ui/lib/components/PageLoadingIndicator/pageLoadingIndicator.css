.root {
    position: relative;
    overflow: hidden;
    width: 100%;
    height: 2px;
    background: rgb(var(--venia-global-color-gray-100));
}

.root_absolute {
    composes: root;
    position: absolute;
    bottom: 0;
    left: 0;
}

.indicator_off {
    position: absolute;
    top: 0;
    left: 0;
    opacity: 0;
    width: 100%;
    height: 100%;
    background: rgb(var(--venia-global-color-blue-600));
    transform: translateX(-100%);
    transform-origin: left;
    transition: transform 0.25s linear;
}

.indicator_loading {
    composes: indicator_off;
    opacity: 1;
    transform: translateX(-25%);
}

.indicator_done {
    composes: indicator_off;
    opacity: 1;
    transform: translateX(0%);
}
