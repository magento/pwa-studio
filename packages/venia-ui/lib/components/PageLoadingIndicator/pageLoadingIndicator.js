import React from 'react';
import { shape, string, bool } from 'prop-types';
import { useStyle } from '../../classify';
import defaultClasses from './pageLoadingIndicator.module.css';
import usePageLoadingIndicator from '@magento/peregrine/lib/talons/PageLoadingIndicator/usePageLoadingIndicator';

const PageLoadingIndicator = props => {
    const classes = useStyle(defaultClasses, props.classes);
    const { absolute } = props;
    const { isPageLoading, loadingState } = usePageLoadingIndicator();

    if (!isPageLoading && !absolute) {
        return null;
    }

    return (
        <div className={absolute ? classes.root_absolute : classes.root}>
            <div className={classes[`indicator_${loadingState}`]} />
        </div>
    );
};

PageLoadingIndicator.defaultProps = {
    classes: {},
    absolute: false
};

PageLoadingIndicator.propTypes = {
    classes: shape({
        root: string
    }),
    absolute: bool
};

export default PageLoadingIndicator;
