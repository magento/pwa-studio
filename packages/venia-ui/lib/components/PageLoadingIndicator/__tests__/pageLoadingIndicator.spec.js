import React from 'react';
import { act, create } from 'react-test-renderer';
import usePageLoadingIndicator from '@magento/peregrine/lib/talons/PageLoadingIndicator/usePageLoadingIndicator';
import PageLoadingIndicator from '../pageLoadingIndicator';

jest.mock(
    '@magento/peregrine/lib/talons/PageLoadingIndicator/usePageLoadingIndicator'
);

jest.mock('../../../classify');

const givenLoadedPage = () => {
    usePageLoadingIndicator.mockImplementation(() => ({
        isPageLoading: false
    }));
};

const givenPageLoading = () => {
    usePageLoadingIndicator.mockImplementation(() => ({
        isPageLoading: true
    }));
};

beforeEach(() => {
    usePageLoadingIndicator.mockClear();
    givenLoadedPage();
});

describe('#PageLoadingIndicator does not render', () => {
    test('when page is not loading', async () => {
        let tree;

        await act(() => {
            tree = create(<PageLoadingIndicator />);
        });

        expect(tree.toJSON()).toBeNull();
    });
});

describe('#PageLoadingIndicator displays', () => {
    test('when page is loading', async () => {
        givenPageLoading();
        let tree;

        await act(() => {
            tree = create(<PageLoadingIndicator />);
        });

        expect(tree.toJSON()).not.toBeNull();
    });

    test('when absolute is true and page is not loading', async () => {
        givenLoadedPage();
        let tree;

        await act(() => {
            tree = create(<PageLoadingIndicator absolute />);
        });

        expect(tree.toJSON()).not.toBeNull();
    });
});
