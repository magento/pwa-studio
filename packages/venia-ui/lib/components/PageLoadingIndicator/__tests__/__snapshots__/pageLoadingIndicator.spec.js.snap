// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#PageLoadingIndicator displays correctly depend on the position type when absolute is false and page is loading 1`] = `
<div
  className="root_relative"
>
  <div
    className="indicator_undefined"
  />
</div>
`;

exports[`#PageLoadingIndicator displays correctly depend on the position type when absolute is true and page is not loading 1`] = `
<div
  className="root_absolute"
>
  <div
    className="indicator_undefined"
  />
</div>
`;
